rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code, prelude::*};
+use core::{marker::PhantomData, mem::ManuallyDrop};
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+pub struct Cpu(i32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: ManuallyDrop<F>,
```

Comment:
I don't think `ManuallyDrop` is the right type for this because it allows undefined behaviour without unsafe block in the way you use it later. In particular, you call `ManuallyDrop::take` in `call_fn` but in `Cpu::call` one could mistakenly try to use after `fn_call.func` after `smp_call_function_single` returns.

I think it would be better to use `MaybeUninit` here; then uses of `func` would require `unsafe` block and we can justify why it's safe at each usage.

Reply:
For code like this, the typical way is to have a union with `ManuallyDrop<F>` and `ManuallyDrop<R>`, see https://doc.rust-lang.org/1.62.1/src/std/panicking.rs.html#413-417. However, I think doing so is premature optimisation; let's use `Option` for now.

Reply:
`Option` it is 👍🏻